File: python/121_best-time-to-buy-and-sell-stock.py
# Dynamic Programming
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 0:
            return 0

        dp = [(0, 0)] * len(prices)  # (minimum_price, profit)
        dp[0] = (prices[0], 0)

        for i in range(1, len(dp)):
            dp[i] = (min(dp[i - 1][0], prices[i]), max(dp[i - 1][1], prices[i] - dp[i - 1][0]))

        return dp[-1][1]


#---------------------------------------------------------------------------------------#


# One Pass
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l = len(prices)
        if l == 0:
            return 0

        minimum_price = prices[0]
        profit = 0

        for i in range(1, l):
            minimum_price = min(minimum_price, prices[i])
            profit = max(profit, prices[i] - minimum_price)
        return profit

File: go/121_best-time-to-buy-and-sell-stock.go
// One Pass
func maxProfit(prices []int) int {
	min, profit := prices[0], 0

	for _, p := range prices[1:] {
		if p < min {
			min = p
		}
		if p-min > profit {
			profit = p - min
		}
	}
	return profit
}

/*---------------------------------------------------------------------------------------*/

// Dynamic Programming
func maxProfit(prices []int) int {
	dp := [][]int{}
	dp = append(dp, []int{prices[0], 0})

	for i, p := range prices[1:] {
		dp = append(dp, []int{min(p, dp[i][0]), max(p-dp[i][0], dp[i][1])})
	}
	return dp[len(dp)-1][1]
}

func min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}
func max(a, b int) int {
	if a < b {
		return b
	} else {
		return a
	}
}